esx_service/utils/auth_api.py

```diff
@@ -350,41 +410,80 @@ def _tenant_vm_ls(name):
         return error_info, None
     
     if not tenant:
-        error_info = error_code.TENANT_NOT_EXIST.format(name)
+        error_info = error_code.generate_error_info(ErrorCode.TENANT_NOT_EXIST, name)              
         return error_info, None
     # tenant.vms is a list of vm_uuid of vms which belong to this tenant
     return None, tenant.vms
+
+def _tenant_vm_replace(name, vm_list):
+    """ API to replace vms for a tenant """
+    logging.debug("_tenant_vm_replace: name=%s vm_list=%s", name, vm_list)
+    if not vm_list:
+        err_code = ErrorCode.REPLACE_VM_EMPTY
+        err_msg = error_code.error_code_to_message[err_code]
+        error_info = ErrorInfo(err_code, err_msg)
+        return error_info
+
+    error_info, tenant = get_tenant_from_db(name)
+    if error_info:
+        return error_info
     
-def _tenant_access_add(name, datastore, allow_create=None, default_datastore=False, volume_maxsize=None, volume_totalsize=None):
+    if not tenant:
+        error_info = error_code.generate_error_info(ErrorCode.TENANT_NOT_EXIST, name)              
+        return error_info
+
+    error_msg, vms, not_found_vms = generate_tuple_from_vm_list(vm_list)
+    
+    if error_msg:
+        not_found_vm_list = ",".join(not_found_vms)
+        error_info = error_code.generate_error_info(ErrorCode.VM_NOT_FOUND, not_found_vm_list)
+        return error_info   
+
+    logging.debug("_tenant_vm_replace: vms=%s", vms)
+    error_info, auth_mgr = get_auth_mgr()
+    if error_info:
```

Comment:
No. Only the "tenant.replace_vms" need auth_mgr.conn. We should not move it up front. And more it is consistent with other similar API like tenant_vm_add, tenant_vm_rm.